drv/cosmo-seq-server/src/main.rs

```diff
@@ -0,0 +1,595 @@
+    ) -> Result<PowerState, RequestError<core::convert::Infallible>> {
+        Ok(self.get_state_impl())
+    }
+
+    fn set_state(
+        &mut self,
+        _: &RecvMessage,
+        state: PowerState,
+    ) -> Result<(), RequestError<CpuSeqError>> {
+        self.set_state_impl(state, StateChangeReason::Other)?;
+        Ok(())
+    }
+
+    fn set_state_with_reason(
+        &mut self,
+        _: &RecvMessage,
+        state: PowerState,
+        reason: StateChangeReason,
+    ) -> Result<(), RequestError<CpuSeqError>> {
+        self.set_state_impl(state, reason)?;
+        Ok(())
+    }
+
+    fn send_hardware_nmi(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<(), RequestError<core::convert::Infallible>> {
+        // The required length for an NMI pulse is apparently not documented.
+        //
+        // Let's try 25 ms!
+        self.sys.gpio_reset(SP_TO_SP5_NMI_SYNC_FLOOD_L);
+        hl::sleep_for(25);
+        self.sys.gpio_set(SP_TO_SP5_NMI_SYNC_FLOOD_L);
+        Ok(())
+    }
+
+    fn read_fpga_regs(
+        &mut self,
+        _: &RecvMessage,
+    ) -> Result<[u8; 64], RequestError<core::convert::Infallible>> {
+        // XXX todo
+        Ok([0; 64])
+    }
+}
+
+impl NotificationHandler for ServerImpl {
+    fn current_notification_mask(&self) -> u32 {
+        notifications::TIMER_MASK
+    }
+
+    fn handle_notification(&mut self, bits: u32) {
+        if (bits & notifications::TIMER_MASK) == 0 {
+            return;
+        }
+        let state = self.log_state_registers();
+        use fmc_periph::{A0Sm, NicSm};
+
+        // Detect unexpected power-off
+        match (self.state, state.seq) {
+            (PowerState::A0 | PowerState::A0PlusHP, Ok(A0Sm::Done)) => (),
+            (PowerState::A0 | PowerState::A0PlusHP, seq_state) => {
+                ringbuf_entry!(Trace::UnexpectedPowerOff {
+                    our_state: self.state,
+                    seq_state,
+                });
+                self.log_pg_registers();
+
+                // Power down to A2, updating our internal state.  We can't
+                // handle errors here, so log them and continue.
+                if let Err(e) = self
+                    .set_state_impl(PowerState::A2, StateChangeReason::Other)
+                {
+                    ringbuf_entry!(Trace::PowerDownError(e))
+                }
+            }
+            // TODO are there other states that we should check here?
+            _ => (),
+        }
+
+        // Detect when the NIC comes online
+        match (self.state, state.nic) {
+            (PowerState::A0, Ok(NicSm::Done)) => {
+                self.set_state_impl(
+                    PowerState::A0PlusHP,
+                    StateChangeReason::InitialPowerOn,
+                )
+                .unwrap(); // this should be infallible
+            }
+            // TODO: should we handle the NIC powering down while the main CPU
+            // power remains up?
```

Comment:
TODO?

Reply:
Let's leave this TODO for now, dunno how we should handle that case (if it should ever occur).